presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```diff
@@ -0,0 +1,205 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.google.sheets;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.sheets.v4.Sheets;
+import com.google.api.services.sheets.v4.SheetsScopes;
+import com.google.api.services.sheets.v4.model.ValueRange;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.LoadingCache;
+import io.airlift.log.Logger;
+import io.prestosql.spi.PrestoException;
+import org.apache.http.annotation.ThreadSafe;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static com.google.common.cache.CacheLoader.from;
+import static com.google.common.util.concurrent.MoreExecutors.newDirectExecutorService;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_BAD_CREDENTIALS_ERROR;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_METASTORE_ERROR;
+import static io.prestosql.plugin.google.sheets.SheetsErrorCode.SHEETS_UNKNOWN_TABLE_ERROR;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@ThreadSafe
+public class GSheetsDataProvider
+        implements SheetsDataProvider
+{
+    private static final Logger log = Logger.get(GSheetsDataProvider.class);
+    private static final String APPLICATION_NAME = "presto google sheets integration";
+    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+    private static final List<String> SCOPES =
+            Collections.singletonList(SheetsScopes.SPREADSHEETS_READONLY);
+
+    private final SheetsConfig sheetsConfig;
+    // tableName <-> sheetId mapping
+    private final LoadingCache<String, Optional<String>> tableSheetMappingCache;
+
+    // sheetId <-> Row List mapping
+    private final LoadingCache<String, Optional<List<List<Object>>>> sheetDataCache;
+
+    private Credential credential;
+    private NetHttpTransport httpTransport;
+    private Sheets sheetsService;
+
+    public GSheetsDataProvider(SheetsConfig config)
+    {
+        this.sheetsConfig = config;
```

Comment:
we typically do not store entire config as fields, but take only required parts out of it. For example please see `io.prestosql.plugin.hive.HiveSplitManager#HiveSplitManager`